File: nodecore-p2p/src/main/java/nodecore/p2p/MessageIdGenerator.java
// VeriBlock NodeCore
// Copyright 2017-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package nodecore.p2p;

import java.util.concurrent.atomic.AtomicLong;

public class MessageIdGenerator {
    private static AtomicLong identity = new AtomicLong(0);

    public static String next() {
        return Long.toString(identity.incrementAndGet());
    }
}

File: nodecore-p2p/src/main/kotlin/nodecore/p2p/MessageIdGenerator.kt
// VeriBlock NodeCore
// Copyright 2017-2021 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package nodecore.p2p

import nodecore.api.grpc.RpcEvent
import java.util.concurrent.atomic.AtomicLong

private val identity = AtomicLong(0)

fun nextMessageId(): String {
    return identity.incrementAndGet().toString()
}

inline fun buildMessage(
    id: String = nextMessageId(),
    acknowledge: Boolean = false,
    buildBlock: RpcEvent.Builder.() -> Unit
): RpcEvent = RpcEvent.newBuilder()
    .setId(id)
    .setAcknowledge(acknowledge)
    .apply(buildBlock)
    .build()
